specs/governance/governor.md

```diff
@@ -0,0 +1,702 @@
+address public manager;
+
+// The token to use for voting.
+IVotesUpgradeable public token;
+
+// Total number of `votes` that `account` has casted for `proposalId`.
+mapping(uint256 proposalId => mapping(address account => uint256 votes)) public weightCast;
+
+// Whether a `module` has been approved or not.
+mapping(address module => bool approved) public approvedModules;
+
+// The vote accounting for a proposal.
+mapping(uint256 proposalId => ProposalVote votes) internal _proposalVotes;
+
+// The proposals that have been created.
+mapping(uint256 proposalId => ProposalCore proposal) internal _proposals;
+```
+
+## Types
+
+The `Governor` contract MUST define the following types:
+
+### `ProposalVote`
+
+`ProposalVote` defines the vote accounting for a proposal, encapsulated in the following struct:
+
+```solidity
+struct ProposalVote {
+  uint256 againstVotes;
+  uint256 forVotes;
+  uint256 abstainVotes;
+  mapping(address => bool) hasVoted;
+}
+```
+
+| Name                     | Type            | Description                                                             |
+|--------------------------|-----------------|-------------------------------------------------------------------------|
+| `againstVotes`          | `uint256` | Amount of votes against the proposal.                         |
+| `forVotes`              | `uint256`       | Amount of votes in favor of the proposal.                |
+| `abstainVotes`                 | `uint256`       | Amount of votes abstaining.   |
+| `hasVoted`                 | `mapping`       | Whether an account has voted for this proposal or not.   |
+
+### `ProposalCore`
+
+`ProposalCore` defines the core information of a proposal, encapsulated in the following struct:
+
+```solidity
+struct ProposalCore {
+  uint64 voteStart;
+  uint64 voteEnd;
+  bool executed;
+  bool canceled;
+  address votingModule;
+  uint8 proposalType;
+}
+```
+
+| Name                     | Type            | Description                                                             |
+|--------------------------|-----------------|-------------------------------------------------------------------------|
+| `voteStart`          | `uint64` | The block timestamp when voting started.                         |
+| `voteEnd`              | `uint64`       | The block timestamp when voting ends.                |
+| `executed`                 | `bool`       | Indicator for proposal execution.   |
+| `canceled`                 | `bool`       | Indicator for proposal cancellation.   |
+| `votingModule`                 | `address`       | The voting module of the proposal. Zero address if none.   |
+| `proposalType`                 | `uint8`       | The type of proposal.   |
+
+## Proposal Types
+
+The `Governor` contract uses proposal types to define the quorum and approval threshold for a proposal to pass.
+These proposal types are created and stored as part of an external contract called `ProposalTypesConfigurator`.
```

Comment:
We should probably include some additional information about this external contract

Reply:
included more details and the interface that the `ProposalTypesConfigurator` must follow so that the governor can use it: https://github.com/ethereum-optimism/specs/pull/315/commits/495398c73b5e5c4825f64b99864e99976a69f673